README
		      The Algol-68RS Compiler (a68toc)
		      --------------------------------

This is the A68toC Algol 68 development system. It was extracted from
the ELLA package which is used for VLSI chip design.

This repository contains a version of a68toc derived from a Linux
port of a68toc by Sian Mountbatten. See README.Sian for more.

The compiler front-end is essentially the Algol-68RS portable
compiler originally written by the Defence Research Agency when it
was known as the RSRE (Royal Signals and Radar Establishment). Only
that part of Algol-68 needed for the ELLA compiler was implemented by
the `a68toc' translator with the result that formats are unavailable
and are likely to remain unavailable for the foreseeable future. The
`PAR' construct, for parallel programming, with the associated mode
`SEMA' and the operators `LEVEL', `UP' and `DOWN' never were part of
the Algol 68 RS compiler.

After extracting the A68toC system from the ELLA package, it became
clear that the library preludes provided by the ELLA team were
entirely unsatisfactory for the general user, not least because they
were almost completely without documentation (see ctrans.texi).

The QAD Standard Prelude given here (in sub-directory qad)
uses the conversion routines from the van Vliet Transput Model of
transput.

This version of the translator dispenses with the module manager `mm' and,
instead, uses `ca' to create preludes and programs.

Starting from Dr Mountbatten's Linux port (labelled as v1.15) this
package has been enhanced with bug-fixes and has machine-specific
dependencies resolved. The result is a sequence of new versions,
starting at v1.20 (a v1.19 version was previously known to exist
but the author has not been able to locate it).

The current branch of a68toc now compiles, runs, and self-compiles
on little-endian CPUs (Intel/AMD) and big-endian CPUs (ARM) in
both 32-bit and 64-bit forms on the Linux operating system
and on Intel 64-bit on macOS.

Support for macOS has been made possible by the integration of the
Boehm-Demers-Weiser Garbage Collector (bdwgc), a version of which
is contained in this repository, and is a compile-time option for
the Linux variants, being set as the default for all 64-bit systems.

Further example programs have been added to this repository including
some from Glyn Webster; see the examples folder.

Neil Matthew
neil@tilde.co.uk
May 2021




debian/changelog
algol68toc (1.21) unstable; urgency=low

  * Ported to ARM 64-bit architecture (aarch64)
  * Ported to Intel 64-bit architecture (x86_64)
  * Included optional Boehm garbage collector (default for 64-bit systems)
  * Added initial support for macOS (darwin)

 -- Neil Matthew <neil@tilde.co.uk>  Mon, 03 Apr 2021 00:00:00 +0100

algol68toc (1.20) unstable; urgency=low

  * Completed addition of envp to main in liba68prel/osshell.a68 and src/a68toc.a68
  * Corrected sprintf formats for LONG INTs in liba68prel/usefulops.a68
  * Ported to ARM 32-bit architecture (armhf)
  * Compile with gcc -fpack-struct=4 on armhf to avoid conformity clause errors on union copies
  * Include <unistd.h> in library/Aerror.h to avoid warning about write()
  * Fix Makefiles to remove a68toc.o on target clean and intall info files correctly

 -- Neil Matthew <neil@tilde.co.uk>  Tue, 13 Apr 2021 15:44:00 +0100

algol68toc (1.15) unstable; urgency=low

  * Added pame to the package and renamed reset.a68 to resetseed

 -- Sian Mountbatten <poenikatu@fastmail.co.uk>  Sun, 30 Sep 2012 19:28:14 +0100

algol68toc (1.14) unstable; urgency=low

  * Ensured that the manuals are included in the package

 -- Sian Mountbatten <poenikatu@fastmail.co.uk>  Fri, 09 Mar 2012 18:55:00 +0000

algol68toc (1.13) unstable; urgency=low

  * Altered ENTIER and ROUND to clamp their input value

 -- Sian Mountbatten <poenikatu@fastmail.co.uk>  Fri, 01 Mar 2012 18:53:35 +0000

algol68toc (1.11) unstable; urgency=low

  * Replaced ROUND and ENTIER with new C functions

 -- Sian Mountbatten <poenikatu@fastmail.co.uk>  Wed, 11 Jan 2012 05:46:54 +0000

algol68toc (1.10) unstable; urgency=low

  * 1. Altered the `new filebuffer' to (INT mtu)FILEBUFFER:
  * 2. Altered bufprocs and `client socket estab' and `server socket estab'
  *    to accept a `c' parameter specifying the MTU. Also altered `accept'

 -- Sian Leitch <drleitch@nildram.co.uk>  Fri, 28 Jan 2005 20:17:16 +0000

algol68toc (1.9) unstable; urgency=low

  * Removed bin possible from `stand in' and `stand out'

 -- Sian Leitch <sian@Phoenicia>  Fri,  6 Aug 2004 23:51:59 +0100

algol68toc (1.8) unstable; urgency=low

  * Added nulls to the id's used by `rename' in `transput.a68'

 -- Sian Leitch <sian@Phoenicia.PhoenixNet>  Sun,  9 Nov 2003 14:34:17 +0000

algol68toc (1.7) unstable; urgency=low

  * Added envp as well as argv

 -- Sian Leitch <algol68@sleitch.nildram.co.uk>  Tue, 10 Jun 2003 14:28:25 +0100

algol68toc (1.6) unstable; urgency=low

  * fixed two bugs in search for rounding point

 -- Sian Leitch <sian@Phoenicia.PhoenixNet>  Fri, 30 May 2003 22:05:15 +0100

algol68toc (1.5) unstable; urgency=low

  * Maintenance release
  * Added to the identifications of the standard files
  * Changed `stand err' to `stand error'
  * Deleted the code to open `stand back'. `stand back' is available, but left
  * unopened
  * Squashed some minor bugs in PAME

 -- Sian Leitch <sian@sleitch.nildram.co.uk>  Wed,  5 Mar 2003 17:24:22 +0000

algol68toc (1.4) unstable; urgency=low

  * Maintenance release
  * Squashed a bug in `concat'

 -- Sian Leitch <sian@sleitch.nildram.co.uk>  Mon,  3 Mar 2003 17:38:08 +0000

algol68toc (1.3) unstable; urgency=low

  * Maintenance release
  * Added a null character (REPR 0) to the end of the command line

 -- Sian Leitch <sian@sleitch.nildram.co.uk>  Tue, 25 Feb 2003 08:35:30 +0000

algol68toc (1.2) unstable; urgency=low

  * Maintenance release
  * Removed all the padding in structures
  * Made argc and argv available as Algol 68 argc and argv
  * Added `ca68' because `make' is not interactive
  * Replaced references to Ctrans in the book by references to Algol68toC
  * and a68toc and corrected minor formatting errors

 -- Dr Sian Leitch (Algol 68 specialist) <algol68@sleitch.nildram.co.uk>  Wed, 19 Jun 2002 22:30:00 +0100

algol68toc (1.1) unstable; urgency=low

  * Maintenance release
  * Changed the name of the package to algol68toc
  * Deleted all references to `mm'
  * Changed directories: stanprel->qad, ctrans->src, amalgamated liba68prel
  * Changed all the Makefiles to remove autobook influence: no configure now

 -- Sian Leitch <algol68@sleitch.nildram.co.uk>  Sat, 23 Feb 2002 17:25:15 +0000

ctrans (1.0.5) unstable; urgency=low

  * Maintenance release
  * Added all the standard prelude identifiers to the index in PAME plus any
  * required text in chapter 13
  * Minor changes to the QAD standard prelude itself
  * Revamped all the Makefiles using autoconf and automake

 -- Sian Leitch <algol68@sleitch.nildram.co.uk>  Wed, 15 Aug 2001 10:57:38 +0100

ctrans (1.0.4) unstable; urgency=low

  * Maintenance release
  * Modified `kbd channel' to avoid loss of input due to the multi-tasking
  * of Linux

 -- Sian Leitch <algol68@sleitch.nildram.co.uk>  Mon, 26 Feb 2001 15:29:48 +0000

ctrans (1.0.3) unstable; urgency=low

  * Maintenance release
  * Added kbd ensure buf space for the kbd channel

 -- Sian Leitch <algol68@sleitch.nildram.co.uk>  Sat, 3 Feb 2001 20:23:00 +0000

ctrans (1.0.2) unstable; urgency=low

  * Maintenance release
  * Kept nearly all the operators from spops.a68
  * Altered the Makefile so that the debug versions would be produced again.

 -- Sian Leitch <algol68@sleitch.nildram.co.uk>  Sun, 12 Nov 2000 20:36:45 +0000

ctrans (1.0.1) unstable; urgency=low

  * Maintenance release
  * Replaced manage/mm_default_database.mmd with the version for the
  * package.
  * Replaced all the `Makefile's with a single top-level Makefile.
  * Recreated all the C source files from the Algol 68 source.
  * Created the necessary `mm' initialisation files so that the Algol
  * 68 source could be translated directly using `mm' and `a68toc'.
  * Cured yet more formatting errors in PAME.
  * Added indexes to the Ctrans and mm info files.
  * Debianised the package.

 -- Sian Leitch <algol68@sleitch.nildram.co.uk>  Wed, 6 Sep 2000 16:23:30 +0100

ctrans (1.0.0) unstable; urgency=low

  * Initial Release.


src/incid.c
/* UNAME:AAAACTR */
#include <algol68/Asupport.h>

A_PROCEDURE(A68_VOID ,A68t31,(A68_BOOL ,A68_VC *),(A68_BOOL ,A68_VC *,void *));
typedef struct A68t31  A68_31 ;    /* PROC(BOOL) MODE26 */

A_PROCEDURE(A68_VOID ,A68t32,(A68_VC ,A68_VC *),(A68_VC ,A68_VC *,void *));
typedef struct A68t32  A68_32 ;    /* PROC(MODE26) REF MODE26 */

A_PROCEDURE(A68_VOID ,A68t33,(void),(void *));
typedef struct A68t33  A68_33 ;    /* PROC VOID */

A_PROCEDURE(A68_VOID ,A68t34,(struct A68t33 ),(struct A68t33 ,void *));
typedef struct A68t34  A68_34 ;    /* PROC(MODE33) VOID */

A_PROCEDURE(A68_VOID ,A68t35,(A68_VC ),(A68_VC ,void *));
typedef struct A68t35  A68_35 ;    /* PROC(MODE26) VOID */

A_PROCEDURE(A68_VOID ,A68t36,(struct A68t35 ),(struct A68t35 ,void *));
typedef struct A68t36  A68_36 ;    /* PROC(MODE35) VOID */

A_PROCEDURE(A68_VOID ,A68t37,(A68_VC ,A68_VC ),(A68_VC ,A68_VC ,void *));
typedef struct A68t37  A68_37 ;    /* PROC(MODE26,MODE26) VOID */
A_ISTRUCT(A68_CHAR ,55,A68t38);
typedef struct A68t38  A68_38 ;    /* STRUCT 55 CHAR */

/* --- Imports from a68config --- */
extern A68_VOID  UEAALIB_a68config(A68_VC ,A68_VC );
static A68_VC  A_configinfo;
/* A cheat - pass a REF through the DECS Module restriction via STR */
#define ACONFIG ( A_configinfo.data = (A68_CHAR*)(&A_config), \
                  A_configinfo.upb = (A68_INT)(-777), A_configinfo )

#define LGAALIB_configinfo ACONFIG
/* --- End of imports from a68config --- */


/* --- DECS initialisation functions --- */
extern void BAAALIB(void);   /* a68config */
/* --- end of DECS initialisation functions --- */
static A68_38   EAAACTR = {"$Id: incid.a68,v 1.1.1.1 2001-05-07 10:16:10 sian Exp $"}; 
A_GISVEC(A68_VC ,FAAACTR,EAAACTR,55)
#define GAAACTR_globalidno 2
#define HAAACTR_minidno 3
#define IAAACTR_maxidno 2000
#define JAAACTR_minrdno 2001
#define KAAACTR_maxrdno 4000
#define LAAACTR_minlibid 4001
#define MAAACTR_maxlibid 6000
#define NAAACTR_keptgeneratorproccorrection 6000
#define OAAACTR_minlab 16384
A68_INT  PAAACTR_maxlab;
#define QAAACTR_minopidno 32768
#define RAAACTR_maxopidno 65536
A68_INT  SAAACTR_minkeptgeneratorproc;
A68_INT  TAAACTR_maxkeptgeneratorproc;
 /* line 1: */
 /* line 3: */
void BAAACTR(void)   /* initialise DECS incid */
{ 
static A68_BOOL A_invoked = A68_FALSE;

/* --- Configuration information for this module */
static char *A_config_arguments[] = {"/home/neil/Algol-68RS/algol68toc-1.21/src/a68toc","-v","-uname","seedfile","-staredit","59LR","-lib","/home/neil/Algol-68RS/algol68toc-1.21/a68config","-dir","/home/neil/Algol-68RS/algol68toc-1.21/a68config","-dir","/home/neil/Algol-68RS/algol68toc-1.21/liba68prel","-dir",".","incid.a68",""};
static char *A_config_environment[] = {"A68_LIB=/home/neil/Algol-68RS/algol68toc-1.21/a68config","A68_CDIR=","A68_DIR=","A68_STAREDIT=","A68_NAMESEED=nameseed","CTRANS_NAMESEED=",""};
static char *A_config_modinfo_files[] = {"/home/neil/Algol-68RS/algol68toc-1.21/a68config/a68config.m",""};
static A_CONFIG_INFO A_config;
/* --- end of configuration information */
if( A_invoked ) return;
A_invoked = A68_TRUE;
BAAALIB();   /* USE a68config */
/* --- Initialise configuration information */
A_config.source_file = "/home/neil/Algol-68RS/algol68toc-1.21/src/incid.a68";
A_config.translation_time = "Wed May  5 17:38:59 2021";
A_config.ctrans_version = "Ctrans_34.6";
A_config.name_seed = "AAAACTR (from seed file) ";
A_config.spec_change_time = "Wed May  5 17:38:59 2021";
A_config.arguments = A_config_arguments;
A_config.environment = A_config_environment;
A_config.modinfo_files = A_config_modinfo_files;
A_PROC_ENTRY(DECS incid);
UEAALIB_a68config(LGAALIB_configinfo, FAAACTR);
 /* line 42: */
 /* line 43: */
 /* line 44: */
 /* line 45: */
 /* line 46: */
 /* line 47: */
 /* line 48: */
 /* line 49: */
 /* line 50: */
 /* line 51: */
PAAACTR_maxlab = (16384+1024);
 /* line 52: */
 /* line 53: */
 /* line 55: */
SAAACTR_minkeptgeneratorproc = (HAAACTR_minidno+NAAACTR_keptgeneratorproccorrection);
 /* line 56: */
 /* line 57: */
TAAACTR_maxkeptgeneratorproc = (IAAACTR_maxidno+NAAACTR_keptgeneratorproccorrection);
 /* line 62: */
A_PROC_EXIT(DECS incid);
} 
#undef NL
/* end of translation of incid.a68 */
